keps/sig-windows/windows-runtimeclass-support.md

```diff
@@ -0,0 +1,575 @@
+- Address feedback on usage/changed behavior, provided on GitHub issues
+- Deprecate the flag
+
+**For non-optional features moving to GA, the graduation criteria must include [conformance tests].**
+
+[conformance tests]: https://github.com/kubernetes/community/blob/master/contributors/devel/conformance-tests.md
+
+### Upgrade / Downgrade Strategy
+
+The new label `node.kubernetes.io/windows-build` can be set or removed if needed without impacting other components as described in [Risks and Mitigations](#risks-and-mitigations)
+
+Users can only opt-in to use the new `runtime_handler` field after setting up and configuring ContainerD. On existing clusters without ContainerD set up, they must use `docker` as the `runtimeHandler` in the `RuntimeClass` today. Therefore they must update to a supported version of ContainerD as a prerequisite which is covered in the scope of another KEP - [Windows CRI-ContainerD].
+
+## Implementation History
+
+Major milestones in the life cycle of a KEP should be tracked in `Implementation History`.
+Major milestones might include
+
+- the `Summary` and `Motivation` sections being merged signaling SIG acceptance
+- the `Proposal` section being merged signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+
+## Alternatives
+
+### Support multiarch os/arch/version in CRI
+
+The Open Container Initiative specifications for container runtime support specifying the architecture, os, and version when pulling and starting a container. This is important for Windows because there is no kernel compatibility between major versions. In order to successfully start a container with process isolation, the right `os.version` must be pulled and run. Hyper-V can provide backwards compatibility, but the image pull and sandbox creation need to specify `os.version` because the kernel is brought up when the sandbox is created. The same challenges exist for Linux as well because multiple CPU architectures can be supported - for example armv7 with qemu and binfmt_misc.
+
+One way to make the experience uniform for dealing with multi-arch images is to add new optional fields to force a deployment to use a specific os/version/arch. This may be combined with RuntimeClass to simplify node placement if needed.
+
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: mypod
+spec:
+  os: windows
+  osversion: 1809
+  architecture: amd64
+  runtimeClassName: windows-hyperv
+  containers:
+      - name: iis
+        image: mcr.microsoft.com/windows/servercore/iis:windowsservercore-ltsc2019
+```
+
+
+Here's the steps needed to pull and start a pod+container matching a specific os/arch/version:
+
+- ImageService.PullImage(PullImageRequest) - PullImageRequest includes a `PodSandboxConfig` reference
```

Comment:
Any idea what `SandboxConfig` is passed here? Is the order for Kubelet

1. `RunPodSandbox`
2. `PullImage(worker1 image, config from step 1)`
3. `CreateContainer(worker1, config from step 1)`
?
Or is PullImage allowed to run before a `RunPodSandbox`? The reason I ask is because, if it is a requirement that `PullImage` happens AFTER `RunPodSandbox` then I guess we can lookup the sandbox `RuntimeHandler` by finding the existing `sandbox` from the `SandboxConfig.Metadata`. This although seems to be a very light contract and I would prefer we dont do it this way.

Reply:
during the main sync loop, sandbox is created first https://github.com/kubernetes/kubernetes/blob/344054bf4aa47fbcf1f3776a8210b7992ea08d11/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L631

then the image is checked, pulled if needed
https://github.com/kubernetes/kubernetes/blob/344054bf4aa47fbcf1f3776a8210b7992ea08d11/pkg/kubelet/kuberuntime/kuberuntime_container.go#L93

then container created then started
https://github.com/kubernetes/kubernetes/blob/344054bf4aa47fbcf1f3776a8210b7992ea08d11/pkg/kubelet/kuberuntime/kuberuntime_container.go#L126


Based on that ordering, I think you're right that the other image actions need the runtimeClass defined.

@Random-Liu - what do you think? Should we

1. keep this as a loose contract as Justin mentioned where the runtimeHandler is assumed to match that of the sandbox

or 

2. also add  `runtimeClass` to all of the image operations: `PullImageRequest` `ListImagesRequest` and `RemoveImageRequest`

Reply:
Yep. The first thing it does is get the `ImageStatus` and use that to determine if a `PullImage` is required which means we do need `RuntimeHandler` for the others.

--

I think that my idea is a bad one for the record. Because it also doesn't solve the `ImageStatus` issue above where it doesnt have the `SandboxConfig`. So even though the sandbox is indeed created we would fail the check.